From 74bbcb4d7d43319ae324e7d25747f4efb58288a9 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Wed, 08 Jun 2022 07:32:04 +0300
Subject: [PATCH] [SW-70527] gaudi2: ras: sram repair support

This patch added support for sram repair flow on sram(48MB).
The flow contains 3 different operations namely
BIRA(Built-in Repair Analysis), BISR(Built-in Soft Repair) and
BIST(Built-in Self Test). Those operations are done sequentially
on 48 SRAM rings.

As BIST operations leaves the memory in dirty state w.r.t ECC,
so all the memory needs to be initialized again. This is done
by scrubbing the memory after BIST.

Change-Id: Ia6c593bec5a8ab117aa60c1fdad0ffafe26ecf39
Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
---

diff --git a/app/ppboot/ppboot.c b/app/ppboot/ppboot.c
index 48f1d1b..e454585 100644
--- a/app/ppboot/ppboot.c
+++ b/app/ppboot/ppboot.c
@@ -272,6 +272,7 @@
 
 __weak void ppboot_security_init(void) {}
 __weak void ppboot_secure_arc(void) {}
+__weak void hl_sram_self_repair(void) {}
 
 void ppboot_main(void)
 {
@@ -308,6 +309,8 @@
 	preboot_delay();
 	sram_set_extension();
 
+	hl_sram_self_repair();
+
 	ret = ppboot_load_preboot_img(&loadaddr);
 	ppboot_check_stack_monitor();
 	if (ret)
diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index adb8216..b150e54 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -9,6 +9,13 @@
 #include <psoc_global_conf_regs.h>
 
 
+#include <psoc_dft_apb_regs.h>
+#include <dcore0_sram0_bank_special_regs.h>
+#include <dcore0_sram0_bank_special_masks.h>
+#include <psoc_global_conf_masks.h>
+#include <cpu_if_regs.h>
+#include <cpu_if_masks.h>
+#include <psoc_reset_conf_regs.h>
 #include <bitops.h>
 #include <preboot.h>
 #include <uart.h>
@@ -18,6 +25,79 @@
 #include <hl.h>
 
 #define SDRAM_BASE_FULL 0x1000FFFFFD000000ull
+
+#define	DELAY_100K_USEC			100000
+#define	DELAY_5K_USEC			5000
+#define	DELAY_67_USEC			67
+
+#define GAUDI2_DFX_NUM			5
+#define GAUDI2_MAX_RING_NUM		28
+
+const static uint8_t dfx_value_table[] = { 0x27, 0x2b, 0x2f, 0x33, 0x37 };
+const static uint8_t ring_count_table[] = { 28, 26, 15, 15, 22 };
+
+const static uint8_t gauid2_ring_type_table[GAUDI2_DFX_NUM][GAUDI2_MAX_RING_NUM] = {
+	{ /* dfx = 0 */
+		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
+		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_SOB_XDMA, SRAM_RING_TYPE_VDEC,
+		SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC,
+		SRAM_RING_TYPE_MME, SRAM_RING_TYPE_HMMU_XBAR, SRAM_RING_TYPE_EDMA1_HMMU3,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_NIC,
+		SRAM_RING_TYPE_HMMU_2, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_XBAR_MID
+	},
+	{ /* dfx = 1 */
+		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
+		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC,
+		SRAM_RING_TYPE_MME, SRAM_RING_TYPE_HMMU_XBAR, SRAM_RING_TYPE_EDMA1_HMMU3,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_XBAR_MID,
+		SRAM_RING_TYPE_HMMU_2, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC,
+		SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX
+	},
+	{ /* dfx = 2 */
+		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
+		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
+		SRAM_RING_TYPE_SOB_ROT, SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC,
+		SRAM_RING_TYPE_HMMU_XBAR, SRAM_RING_TYPE_HMMU_MC, SRAM_RING_TYPE_XBAR_MID,
+		SRAM_RING_TYPE_HMMU_2, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX
+	},
+	{ /* dfx = 3 */
+		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
+		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
+		SRAM_RING_TYPE_SOB_ROT, SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC,
+		SRAM_RING_TYPE_HMMU_XBAR, SRAM_RING_TYPE_HMMU_MC, SRAM_RING_TYPE_XBAR_MID,
+		SRAM_RING_TYPE_HMMU_2, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX
+	}, /* dfx = 4 */
+	{
+		SRAM_RING_TYPE_MME, SRAM_RING_TYPE_MME, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
+		SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
+		SRAM_RING_TYPE_MAX
+	},
+};
 
 static struct hl_pll_cfg gaudi2_arch_pll_cfg[] = {
 	{
@@ -83,3 +163,261 @@
 	for (i = 0; i < (size / 4); i++)
 		*(dst + i) = *(src + i);
 }
+
+static int gaudi2_sms_check_pattern(uint32_t value, uint32_t ones_mask, uint32_t zeros_mask)
+{
+	if ((value & ones_mask) != ones_mask) {
+		preboot_puts("ones check failed\n");
+		return 1;
+	}
+
+	if ((~value & zeros_mask) != zeros_mask) {
+		preboot_puts("zeros check failed\n");
+		return 1;
+	}
+
+	return 0;
+}
+
+static int gaudi2_sms_read_and_check_pattern(uint32_t ones_mask, uint32_t zeros_mask)
+{
+	uint32_t val;
+
+	val = regmap_readl(mmPSOC_DFT_APB_WORD1);
+
+	gaudi2_sms_check_pattern(val, ones_mask, zeros_mask);
+
+	return 0;
+}
+
+static void gaudi2_sram_ring_pattern(enum gaudi2_sram_self_repair_op op)
+{
+	uint32_t expected_ones, expected_zeros;
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
+	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL2, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_5K_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL3, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_67_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL4, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_5K_USEC);
+	} else {
+		preboot_puts("Invalid sram repair operation\n");
+		return;
+	}
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL5, mmPSOC_DFT_APB_WORD0);
+	if (op == SRAM_SELF_REPAIR_OP_BIST)
+		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
+	else
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000000000000000001010000000;
+	expected_zeros = 0b00000000000000000000000101110000;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_5K_USEC);
+
+		regmap_writel(GAUDI2_RING_PATTERN_VAL10, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL11,  mmPSOC_DFT_APB_WORD1);
+
+		expected_ones = 0b00000000000000000000000000000000;
+		expected_zeros = 0b00000000000000000000000000010000;
+		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	}
+}
+
+static void gaudi2_sram_ring_sel(uint32_t dfx_val, uint32_t ring)
+{
+	uint32_t ring_value = BIT(8) | (ring << 3) | 0x5;
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL12, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL13, mmPSOC_DFT_APB_WORD1);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL14, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(dfx_val, mmPSOC_DFT_APB_WORD1);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL12, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL15, mmPSOC_DFT_APB_WORD1);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL10, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL16, mmPSOC_DFT_APB_WORD1);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL10, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(ring_value, mmPSOC_DFT_APB_WORD1);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL10, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL17, mmPSOC_DFT_APB_WORD1);
+}
+
+static void gaudi2_sram_id_sel(void)
+{
+	regmap_writel(GAUDI2_RING_PATTERN_VAL12, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL13, mmPSOC_DFT_APB_WORD1);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL14, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL18, mmPSOC_DFT_APB_WORD1);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL12, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL19, mmPSOC_DFT_APB_WORD1);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL20, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL21, mmPSOC_DFT_APB_WORD1);
+}
+
+static void gaudi2_sram_beginning_pattern(void)
+{
+	regmap_writel(GAUDI2_RING_PATTERN_VAL22, mmPSOC_GLOBAL_CONF_DFT_APB_CONTROL);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL22, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+}
+
+static void (*gaudi2_sram_region_ring_pattern[])(enum gaudi2_sram_self_repair_op op)  = {
+				[SRAM_RING_TYPE_TPC] = NULL,
+				[SRAM_RING_TYPE_SOB_ROT] = NULL,
+				[SRAM_RING_TYPE_SRAM] = gaudi2_sram_ring_pattern,
+				[SRAM_RING_TYPE_XBAR_MID] = NULL,
+				[SRAM_RING_TYPE_VDEC] = NULL,
+				[SRAM_RING_TYPE_SOB_XDMA] = NULL,
+				[SRAM_RING_TYPE_NIC] = NULL,
+				[SRAM_RING_TYPE_MME] = NULL,
+				[SRAM_RING_TYPE_HMMU_XBAR] = NULL,
+				[SRAM_RING_TYPE_HMMU_2] = NULL,
+				[SRAM_RING_TYPE_EDMA1_HMMU3] = NULL,
+				[SRAM_RING_TYPE_HMMU_MC] = NULL
+};
+
+static void gaudi2_sram_self_repair(enum gaudi2_sram_self_repair_op op)
+{
+	unsigned int dfx, ring;
+	enum gaudi2_sram_ring_type ring_type;
+
+	gaudi2_sram_beginning_pattern();
+
+	for (dfx = 0; dfx < ARRAY_SIZE(dfx_value_table); dfx++) {
+		for (ring = 0; ring < ring_count_table[dfx]; ring++) {
+			ring_type = gauid2_ring_type_table[dfx][ring];
+			if (ring_type == SRAM_RING_TYPE_SRAM) {
+				gaudi2_sram_id_sel();
+				gaudi2_sram_ring_sel(dfx_value_table[dfx], ring + 1);
+				gaudi2_sram_region_ring_pattern[ring_type](op);
+				gaudi2_sram_id_sel();
+			}
+		}
+	}
+}
+
+static void gaudi2_mask_sram_ecc(bool enable)
+{
+	u32 val;
+	int dcore, sram;
+	u32 dcore_offset = 0x200000;
+	u32 sram_offset = 0x8000;
+	u32 offset;
+
+	if (enable) {
+		regmap_writel(CPU_IF_SRAM_DERR_INTR_MASK_VAL_MASK, mmCPU_IF_SRAM_DERR_INTR_MASK);
+		regmap_writel(CPU_IF_SRAM_SERR_INTR_MASK_VAL_MASK, mmCPU_IF_SRAM_SERR_INTR_MASK);
+		val = DCORE0_SRAM0_BANK_SPECIAL_MEM_ECC_GLBL_ERR_MASK_DERR_MASK |
+				DCORE0_SRAM0_BANK_SPECIAL_MEM_ECC_GLBL_ERR_MASK_SERR_MASK;
+	} else {
+		regmap_writel(0, mmCPU_IF_SRAM_SERR_INTR_MASK);
+		regmap_writel(0, mmCPU_IF_SRAM_DERR_INTR_MASK);
+		val = 0;
+	}
+
+	for (dcore = 0; dcore < NUM_OF_DCORES ; dcore++) {
+		offset = dcore * dcore_offset;
+		for (sram = 0; sram < NUM_OF_SRAM_BANK_PER_DCORE; sram++) {
+			regmap_writel(val, mmDCORE0_SRAM0_BANK_SPECIAL_MEM_ECC_GLBL_ERR_MASK
+											+ offset);
+			offset += sram_offset;
+		}
+	}
+}
+
+static void gaudi2_mask_clear_sram_ecc(void)
+{
+	int dcore, sram, wrap;
+	u32 dcore_offset = 0x200000;
+	u32 sram_offset = 0x8000;
+	u32 offset;
+	u32 num_mem_wrapper = 0;
+
+	num_mem_wrapper = regmap_readl(mmDCORE0_SRAM0_BANK_SPECIAL_MEM_NUMOF);
+	for (dcore = 0; dcore < NUM_OF_DCORES ; dcore++) {
+		offset = dcore * dcore_offset;
+		for (sram = 0; sram < NUM_OF_SRAM_BANK_PER_DCORE; sram++) {
+			for (wrap = 0; wrap < num_mem_wrapper; wrap++) {
+				regmap_writel(wrap,
+						mmDCORE0_SRAM0_BANK_SPECIAL_MEM_ECC_SEL + offset);
+				regmap_writel((
+					DCORE0_SRAM0_BANK_SPECIAL_MEM_ECC_CTL_DERR_CLR_MASK |
+					DCORE0_SRAM0_BANK_SPECIAL_MEM_ECC_CTL_SERR_CLR_MASK),
+					mmDCORE0_SRAM0_BANK_SPECIAL_MEM_ECC_CTL + offset);
+			}
+			offset += sram_offset;
+		}
+	}
+}
+
+#define SRAM_NUM_MEM_CPY_XFERS			48 /* (48*1024*1024)/(65535*16) */
+static void gaudi2_mask_scrub_sram(void)
+{
+	u32 retry = 1000;
+	u32 mem_cfg, cpy_sts;
+	u32 size = 0xFFFF0; /* (65535*16) bytes */
+	u32 block;
+	u64 addr;
+
+	for (block = 0; block <= SRAM_NUM_MEM_CPY_XFERS; block++) {
+		addr = SDRAM_BASE_FULL + size * block;
+		if (block == SRAM_NUM_MEM_CPY_XFERS) /* last block */
+			size = 768; /* (48*1024*1024) - (48*65535*16) bytes */
+
+		/* Use memblock to initialize dccm memory */
+		regmap_writel((u32)addr, mmPSOC_GLOBAL_CONF_MEM_CPY_DEST_ADDR_L);
+		regmap_writel((u32)(addr >> 32), mmPSOC_GLOBAL_CONF_MEM_CPY_DEST_ADDR_H);
+
+		mem_cfg = regmap_readl(mmPSOC_GLOBAL_CONF_MEM_CPY_CTRL2);
+		mem_cfg &= ~(PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_MEM_SIZE_MASK);
+
+		/* set size of in granuality of 16 bytes */
+		mem_cfg |= ((size >> 4) & PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_MEM_SIZE_MASK);
+		mem_cfg |= BIT(PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_USE_CONST_SHIFT);
+
+		regmap_writel(mem_cfg, mmPSOC_GLOBAL_CONF_MEM_CPY_CTRL2);
+		/* Enable copy and wait for completion */
+		regmap_writel(BIT(0), mmPSOC_GLOBAL_CONF_MEM_CPY_CTRL);
+		while (1) {
+			cpy_sts = regmap_readl(mmPSOC_GLOBAL_CONF_MEM_CPY_STATUS);
+			if (cpy_sts)
+				break;
+
+			retry--;
+			if (!retry) {
+				preboot_puts("sram mem_cpy timedout");
+				break;
+			}
+			udelay(DELAY_100K_USEC);
+		}
+	}
+}
+
+static void gaudi2_scrub_sram(void)
+{
+	gaudi2_mask_sram_ecc(true);
+	gaudi2_mask_scrub_sram();
+	gaudi2_mask_clear_sram_ecc();
+	gaudi2_mask_sram_ecc(false);
+}
+
+#define DFT_APB_CONTROL_DEFAULT_VAL		0xFE
+void hl_sram_self_repair(void)
+{
+	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BIRA);
+	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BISR);
+	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BIST);
+	/* Get back to normal from DFT mode */
+	regmap_writel(DFT_APB_CONTROL_DEFAULT_VAL, mmPSOC_GLOBAL_CONF_DFT_APB_CONTROL);
+	gaudi2_scrub_sram(); /* TODO: Replace this by UNIT reset (Refer SIV-1088) */
+	preboot_puts("sram bist done\n");
+}
diff --git a/arch/gaudi2/common/common.h b/arch/gaudi2/common/common.h
index 732d8a0..ad9112e 100644
--- a/arch/gaudi2/common/common.h
+++ b/arch/gaudi2/common/common.h
@@ -200,6 +200,52 @@
 	TPC_ID_SIZE,
 };
 
+enum gaudi2_sram_self_repair_op {
+	SRAM_SELF_REPAIR_OP_BIRA,
+	SRAM_SELF_REPAIR_OP_BISR,
+	SRAM_SELF_REPAIR_OP_BIST
+};
+
+enum gaudi2_sram_ring_type {
+	SRAM_RING_TYPE_TPC,
+	SRAM_RING_TYPE_SOB_ROT,
+	SRAM_RING_TYPE_SRAM,
+	SRAM_RING_TYPE_XBAR_MID,
+	SRAM_RING_TYPE_VDEC,
+	SRAM_RING_TYPE_SOB_XDMA,
+	SRAM_RING_TYPE_NIC,
+	SRAM_RING_TYPE_MME,
+	SRAM_RING_TYPE_HMMU_XBAR,
+	SRAM_RING_TYPE_HMMU_2,
+	SRAM_RING_TYPE_EDMA1_HMMU3,
+	SRAM_RING_TYPE_HMMU_MC,
+	SRAM_RING_TYPE_MAX
+};
+
+#define GAUDI2_RING_PATTERN_VAL1		0x1000012
+#define GAUDI2_RING_PATTERN_VAL2		0x121
+#define GAUDI2_RING_PATTERN_VAL3		0x131
+#define GAUDI2_RING_PATTERN_VAL4		0x111
+#define GAUDI2_RING_PATTERN_VAL5		0x1000013
+#define GAUDI2_RING_PATTERN_VAL6		0x9
+#define GAUDI2_RING_PATTERN_VAL7		0x7
+#define GAUDI2_RING_PATTERN_VAL8		0x1000012
+#define GAUDI2_RING_PATTERN_VAL9		0x1
+#define GAUDI2_RING_PATTERN_VAL10		0x100000d
+#define GAUDI2_RING_PATTERN_VAL11		0x7
+#define GAUDI2_RING_PATTERN_VAL12		0x1000008
+#define GAUDI2_RING_PATTERN_VAL13		0x85
+#define GAUDI2_RING_PATTERN_VAL14		0x1000006
+#define GAUDI2_RING_PATTERN_VAL15		0xf1
+#define GAUDI2_RING_PATTERN_VAL16		0x109
+#define GAUDI2_RING_PATTERN_VAL17		0x1e1
+#define GAUDI2_RING_PATTERN_VAL18		0x23
+#define GAUDI2_RING_PATTERN_VAL19		0x79
+#define GAUDI2_RING_PATTERN_VAL20		0x100000a
+#define GAUDI2_RING_PATTERN_VAL21		0x3
+#define GAUDI2_RING_PATTERN_VAL22		0x1000001
+
+
 void gaudi2_axi_drain_ena(void);
 uint32_t gaudi2_gpio_read(uint32_t gpio_pin);
 bool preboot_fake_security_is_enabled(void);
diff --git a/include/preboot.h b/include/preboot.h
index 094669d..8f0b300 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -166,6 +166,7 @@
 void ppboot_cp_preboot(u64 loadaddr, u32 offset, u32 size, char *name);
 void image_start(u64 loadaddr);
 u32 sp_sram_readl(u32 offset);
+void hl_sram_self_repair(void);
 
 struct preboot_fuse_ver_data {
 	char *table_ver;
